recent compiler crashes when compiling bun

### Zig Version

0.11.0-dev.4059+17255bed4

### Steps to Reproduce and Observed Behavior

1. `git clone https://github.com/oven-sh/bun`
2. `zig build obj`
3. 💥 segfault

```zig
zig build-obj bun-debug Debug native-native-gnu.2.27: error: the following command terminated unexpectedly:
/home/jarred/zig/0.11.0-dev.4059+17255bed4/files/zig build-obj -freference-trace=256 /home/jarred/Code/bun/root.zig -lc -femit-bin=/home/jarred/Code/bun/packages/debug-bun-linux-x64/bun-debug.o -fno-strip --eh-frame-hdr --emit-relocs -ffunction-sections --cache-dir /home/jarred/Code/bun/zig-cache --global-cache-dir /home/jarred/.cache/zig --name bun-debug -fno-compiler-rt -fno-stack-check -fno-omit-frame-pointer -target native-native-gnu.2.27 -mcpu haswell --mod build_options::/home/jarred/Code/bun/zig-cache/c/152e930bfb125fe9e939c2660f3432b5/options.zig --mod async_io::/home/jarred/Code/bun/src/io/io_linux.zig --deps async_io,build_options -I /home/jarred/Code/bun/src/deps -I /home/jarred/Code/bun/src/deps --main-pkg-path /home/jarred/Code/bun --listen=- 
Build Summary: 1/3 steps succeeded; 1 failed (disable with --summary none)
obj transitive failure
└─ zig build-obj bun-debug Debug native-native-gnu.2.27 failure
error: the following build command failed with exit code 1:
/home/jarred/Code/bun/zig-cache/o/4d1f23298632acbf6ac0a879ce219c2a/build /home/jarred/zig/0.11.0-dev.4059+17255bed4/files/zig /home/jarred/Code/bun /home/jarred/Code/bun/zig-cache /home/jarred/.cache/zig obj -freference-trace -freference-trace -Dcpu=native
make: *** [Makefile:1099: dev-obj-track] Error 1
```

Here is a very unhelpful error stack trace from lldb:
```zig
 thread #1, name = 'zig', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
  * frame #0: 0x0000000007d74e0a zig`___lldb_unnamed_symbol160694 + 58
    frame #1: 0x0000000007ec9098 zig`___lldb_unnamed_symbol161287 + 120
    frame #2: 0x0000000007ec2fc4 zig`___lldb_unnamed_symbol161252 + 4372
    frame #3: 0x0000000007ec471f zig`___lldb_unnamed_symbol161255 + 47
    frame #4: 0x0000000007eb169d zig`___lldb_unnamed_symbol161179 + 605
    frame #5: 0x0000000007ebc973 zig`___lldb_unnamed_symbol161237 + 3075
    frame #6: 0x0000000007e9740b zig`___lldb_unnamed_symbol161172 + 3211
    frame #7: 0x0000000007e8da4e zig`___lldb_unnamed_symbol161154 + 3854
    frame #8: 0x0000000007e7db2f zig`___lldb_unnamed_symbol161135 + 7615
    frame #9: 0x0000000007e7ba22 zig`___lldb_unnamed_symbol161132 + 1010
    frame #10: 0x00000000081d2ca7 zig`___lldb_unnamed_symbol163986 + 1143
    frame #11: 0x00000000081d23ca zig`___lldb_unnamed_symbol163985 + 362
    frame #12: 0x0000000007e32055 zig`___lldb_unnamed_symbol160955 + 15701
    frame #13: 0x00000000081d3505 zig`___lldb_unnamed_symbol163988 + 165
    frame #14: 0x0000000007e34ba8 zig`___lldb_unnamed_symbol160955 + 26792
    frame #15: 0x00000000081d1872 zig`___lldb_unnamed_symbol163983 + 706
    frame #16: 0x0000000007e53547 zig`___lldb_unnamed_symbol160955 + 152135
    frame #17: 0x00000000081d3505 zig`___lldb_unnamed_symbol163988 + 165
    frame #18: 0x0000000007e34ba8 zig`___lldb_unnamed_symbol160955 + 26792
    frame #19: 0x00000000081d3505 zig`___lldb_unnamed_symbol163988 + 165
    frame #20: 0x0000000007e34ba8 zig`___lldb_unnamed_symbol160955 + 26792
    frame #21: 0x00000000081d1872 zig`___lldb_unnamed_symbol163983 + 706
    frame #22: 0x0000000007e53547 zig`___lldb_unnamed_symbol160955 + 152135
    frame #23: 0x00000000081d3505 zig`___lldb_unnamed_symbol163988 + 165
    frame #24: 0x0000000007e34ba8 zig`___lldb_unnamed_symbol160955 + 26792
    frame #25: 0x00000000081d3505 zig`___lldb_unnamed_symbol163988 + 165
    frame #26: 0x0000000007e34ba8 zig`___lldb_unnamed_symbol160955 + 26792
    frame #27: 0x0000000007e7ce35 zig`___lldb_unnamed_symbol161135 + 4293
    frame #28: 0x0000000007e7ba22 zig`___lldb_unnamed_symbol161132 + 1010
    frame #29: 0x00000000081d2ca7 zig`___lldb_unnamed_symbol163986 + 1143
    frame #30: 0x00000000081d23ca zig`___lldb_unnamed_symbol163985 + 362
    frame #31: 0x0000000007e32055 zig`___lldb_unnamed_symbol160955 + 15701
    frame #32: 0x00000000081d3505 zig`___lldb_unnamed_symbol163988 + 165
    frame #33: 0x0000000007e34ba8 zig`___lldb_unnamed_symbol160955 + 26792
    frame #34: 0x00000000081d1bd3 zig`___lldb_unnamed_symbol163983 + 1571
    frame #35: 0x0000000007e53547 zig`___lldb_unnamed_symbol160955 + 152135
    frame #36: 0x00000000081d3505 zig`___lldb_unnamed_symbol163988 + 165
    frame #37: 0x0000000007e34ba8 zig`___lldb_unnamed_symbol160955 + 26792
    frame #38: 0x00000000081d3505 zig`___lldb_unnamed_symbol163988 + 165
    frame #39: 0x0000000007e34ba8 zig`___lldb_unnamed_symbol160955 + 26792
    frame #40: 0x0000000007e7ce35 zig`___lldb_unnamed_symbol161135 + 4293
    frame #41: 0x0000000007d8feac zig`___lldb_unnamed_symbol160769 + 3116
    frame #42: 0x0000000007d7d698 zig`___lldb_unnamed_symbol160713 + 22920
    frame #43: 0x0000000007d6c659 zig`___lldb_unnamed_symbol160675 + 473
    frame #44: 0x0000000007c821d1 zig`___lldb_unnamed_symbol159719 + 455841
    frame #45: 0x0000000007c0ea67 zig`___lldb_unnamed_symbol159703 + 11319
    frame #46: 0x0000000007c0bdc7 zig`___lldb_unnamed_symbol159701 + 199
(lldb) 
```

### Expected Behavior

no segfault. maybe an error message if there was a breaking change between bf827d0b5 and 17255bed4

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -30706,6 +30706,41 @@ fn analyzeIsNonErrComptimeOnly(
     const set_ty = ip.errorUnionSet(operand_ty.toIntern());
     switch (set_ty) {
         .anyerror_type => {},
+        .adhoc_inferred_error_set_type => if (sema.fn_ret_ty_ies) |ies| blk: {
+            // If the error set is empty, we must return a comptime true or false.
+            // However we want to avoid unnecessarily resolving an inferred error set
+            // in case it is already non-empty.
+            switch (ies.resolved) {
+                .anyerror_type => break :blk,
+                .none => {},
+                else => |i| if (ip.indexToKey(i).error_set_type.names.len != 0) break :blk,
+            }
+
+            if (maybe_operand_val != null) break :blk;
+
+            // Try to avoid resolving inferred error set if possible.
+            if (ies.errors.count() != 0) return .none;
+            switch (ies.resolved) {
+                .anyerror_type => return .none,
+                .none => {},
+                else => switch (ip.indexToKey(ies.resolved).error_set_type.names.len) {
+                    0 => return .bool_true,
+                    else => return .none,
+                },
+            }
+            for (ies.inferred_error_sets.keys()) |other_ies_index| {
+                if (set_ty == other_ies_index) continue;
+                const other_resolved =
+                    try sema.resolveInferredErrorSet(block, src, other_ies_index);
+                if (other_resolved == .anyerror_type) {
+                    ies.resolved = .anyerror_type;
+                    return .none;
+                }
+                if (ip.indexToKey(other_resolved).error_set_type.names.len != 0)
+                    return .none;
+            }
+            return .bool_true;
+        },
         else => switch (ip.indexToKey(set_ty)) {
             .error_set_type => |error_set_type| {
                 if (error_set_type.names.len == 0) return .bool_true;
@@ -30719,41 +30754,38 @@ fn analyzeIsNonErrComptimeOnly(
                     .none => {},
                     else => |i| if (ip.indexToKey(i).error_set_type.names.len != 0) break :blk,
                 }
-                if (maybe_operand_val == null) {
-                    if (sema.fn_ret_ty_ies) |ies| {
-                        if (set_ty == .adhoc_inferred_error_set_type or
-                            ies.func == func_index)
-                        {
-                            // Try to avoid resolving inferred error set if possible.
-                            if (ies.errors.count() != 0) return .none;
-                            switch (ies.resolved) {
-                                .anyerror_type => return .none,
-                                .none => {},
-                                else => switch (ip.indexToKey(ies.resolved).error_set_type.names.len) {
-                                    0 => return .bool_true,
-                                    else => return .none,
-                                },
-                            }
-                            for (ies.inferred_error_sets.keys()) |other_ies_index| {
-                                if (set_ty == other_ies_index) continue;
-                                const other_resolved =
-                                    try sema.resolveInferredErrorSet(block, src, other_ies_index);
-                                if (other_resolved == .anyerror_type) {
-                                    ies.resolved = .anyerror_type;
-                                    return .none;
-                                }
-                                if (ip.indexToKey(other_resolved).error_set_type.names.len != 0)
-                                    return .none;
-                            }
-                            return .bool_true;
+                if (maybe_operand_val != null) break :blk;
+                if (sema.fn_ret_ty_ies) |ies| {
+                    if (ies.func == func_index) {
+                        // Try to avoid resolving inferred error set if possible.
+                        if (ies.errors.count() != 0) return .none;
+                        switch (ies.resolved) {
+                            .anyerror_type => return .none,
+                            .none => {},
+                            else => switch (ip.indexToKey(ies.resolved).error_set_type.names.len) {
+                                0 => return .bool_true,
+                                else => return .none,
+                            },
                         }
-                    }
-                    const resolved_ty = try sema.resolveInferredErrorSet(block, src, set_ty);
-                    if (resolved_ty == .anyerror_type)
-                        break :blk;
-                    if (ip.indexToKey(resolved_ty).error_set_type.names.len == 0)
+                        for (ies.inferred_error_sets.keys()) |other_ies_index| {
+                            if (set_ty == other_ies_index) continue;
+                            const other_resolved =
+                                try sema.resolveInferredErrorSet(block, src, other_ies_index);
+                            if (other_resolved == .anyerror_type) {
+                                ies.resolved = .anyerror_type;
+                                return .none;
+                            }
+                            if (ip.indexToKey(other_resolved).error_set_type.names.len != 0)
+                                return .none;
+                        }
                         return .bool_true;
+                    }
                 }
+                const resolved_ty = try sema.resolveInferredErrorSet(block, src, set_ty);
+                if (resolved_ty == .anyerror_type)
+                    break :blk;
+                if (ip.indexToKey(resolved_ty).error_set_type.names.len == 0)
+                    return .bool_true;
             },
             else => unreachable,
         },

--- a/src/type.zig
+++ b/src/type.zig
@@ -2065,7 +2065,7 @@ pub const Type = struct {
     pub fn errorSetIsEmpty(ty: Type, mod: *Module) bool {
         const ip = &mod.intern_pool;
         return switch (ty.toIntern()) {
-            .anyerror_type => false,
+            .anyerror_type, .adhoc_inferred_error_set_type => false,
             else => switch (ip.indexToKey(ty.toIntern())) {
                 .error_set_type => |error_set_type| error_set_type.names.len == 0,
                 .inferred_error_set_type => |i| switch (ip.funcIesResolved(i).*) {
@@ -2084,6 +2084,7 @@ pub const Type = struct {
         const ip = &mod.intern_pool;
         return switch (ty.toIntern()) {
             .anyerror_type => true,
+            .adhoc_inferred_error_set_type => false,
             else => switch (mod.intern_pool.indexToKey(ty.toIntern())) {
                 .inferred_error_set_type => |i| ip.funcIesResolved(i).* == .anyerror_type,
                 else => false,

--- a/src/value.zig
+++ b/src/value.zig
@@ -483,10 +483,10 @@ pub const Value = struct {
     }
 
     pub fn getFunction(val: Value, mod: *Module) ?InternPool.Key.Func {
-        return switch (mod.intern_pool.indexToKey(val.toIntern())) {
+        return if (val.ip_index != .none) switch (mod.intern_pool.indexToKey(val.toIntern())) {
             .func => |x| x,
             else => null,
-        };
+        } else null;
     }
 
     pub fn getExternFunc(val: Value, mod: *Module) ?InternPool.Key.ExternFunc {

